### DISCUSSION TITLE
Split batched big graph into original sub-graph or masking

### DISCUSSION THREAD

**Discussion by JueXiuHuang** [pyg-team/pytorch_geometric#4155] [General]
Hi,
currently I'm working on a Graph2Text task, and I'm trying to do graph-text alignment in my model.
But DataLoader in PyG will stack the batched sub-graphs into a big graph, how can I split them into original sub-graph or using a mask for alignment?

Thank!

**Comment by fengsxy** [comment]
I have the the same  question,i will prepare three  graph together,the model need to handle the three graph.At first , i try to concate them.And then use index to split them.
![image](https://user-images.githubusercontent.com/57928993/155829777-bc1a45cb-9eb4-48a7-b078-fc7800bf92e5.png)
But then because of the stack charaters,the index is over it.
for example my x [9,484]
but edge_index is over 3000
so i want find a way to solve it !

**Reply by rusty1s** [reply]
Yes, you will need to take care of decrementing `edge_index` for each individual example. You can use `SparseTensor` for this (which does it automatically):
```python
from torch_sparse import SparseTensor
adj = SparseTensor.from_edge_index(edge_index, edge_attr)
adj = adj[start_idx:end_idx, start_idx:end_idx]
(row, col), edge_attr = adj.coo()
edge_index = torch.stack([row, col], dim=0)
```
Does that answer your question?

**Reply by fengsxy** [reply]
yeah,Thank u for your answer, i think that may be ok. This time I use the information which is edge length  i stoage in  dataset  to recompute the edge_index,and that's have already been ok.But i still have one question ,if i have three pairs of edge_index,edge_attr,can  i use anything 
 to make three pair of edge indexs stack independtly and automatically.Like the one.Really thanks for your answer.

**Reply by rusty1s** [reply]
You can also use `SparseTensor` for this:
```python
from torch_sparse import cat, SparseTensor

adjs = [SparseTensor.from_edge_index(edge_index) for edge_index in edge_indices]
adj = cat(adjs, dim=(0, 1))  # diagonal stacking
```

**Comment by rusty1s** [comment]
@JueXiuHuang Can you clarify on what you want to do exactly? In general, you can use `batch` to index select the components of an individual graph. Furthermore, the `to_dense_batch` function will convert your node features to shape `[batch_size, num_nodes, num_features]`. Does that help in your case?

**Reply by JueXiuHuang** [reply]
Thanks! This information helps a lot!
